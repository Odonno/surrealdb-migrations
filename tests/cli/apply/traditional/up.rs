use assert_fs::TempDir;
use color_eyre::eyre::{ensure, Result};
use insta::{assert_snapshot, Settings};

use crate::helpers::*;

#[test]
fn apply_initial_migrations() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    println!("{}", db_name);

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply");

    cmd.assert().try_success().and_then(|assert| {
        assert.try_stdout(
            "Executing migration Initial...
Executing migration AddAdminUser...
Executing migration AddPost...
Executing migration CommentPost...
Migration files successfully executed!\n",
        )
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn apply_new_migrations() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;

    let second_migration_name = get_second_migration_name(&temp_dir)?;
    apply_migrations_up_to(&temp_dir, &db_name, &second_migration_name)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply");

    cmd.assert().try_success().and_then(|assert| {
        assert.try_stdout(
            "Executing migration AddPost...
Executing migration CommentPost...
Migration files successfully executed!\n",
        )
    })?;

    temp_dir.close()?;

    Ok(())
}

#[tokio::test]
#[ignore = "should 'dry-run' run against a concrete database?"]
async fn apply_initial_migrations_in_dry_run() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply").arg("--dry-run");

    cmd.assert()
        .try_success()
        .and_then(|assert| assert.try_stdout(""))?;

    let is_empty = is_surreal_db_empty(None, Some(db_name)).await?;
    ensure!(is_empty, "SurrealDB should be empty");

    temp_dir.close()?;

    Ok(())
}

#[test]
fn apply_with_inlined_down_files() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;
    inline_down_migration_files(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply");

    cmd.assert().try_success().and_then(|assert| {
        assert.try_stdout(
            "Executing migration Initial...
Executing migration AddAdminUser...
Executing migration AddPost...
Executing migration CommentPost...
Migration files successfully executed!\n",
        )
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
#[ignore = "should 'dry-run' run against a concrete database?"]
fn apply_and_output_initial_migrations() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply").arg("--dry-run").arg("-o");

    cmd.assert().try_success().map(|assert| {
        let out = String::from_utf8_lossy(&assert.get_output().stdout);

        let mut settings = Settings::new();
        settings.add_script_timestamp_filter();

        settings.bind(|| {
            assert_snapshot!(out);
        });
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn apply_and_output_new_migrations() -> Result<()> {
    let temp_dir = TempDir::new()?;
    let db_name = generate_random_db_name()?;

    add_migration_config_file_with_db_name(&temp_dir, DbInstance::Root, &db_name)?;
    scaffold_blog_template(&temp_dir, true)?;

    let first_migration_name = get_first_migration_name(&temp_dir)?;
    apply_migrations_up_to(&temp_dir, &db_name, &first_migration_name)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.arg("apply").arg("--dry-run").arg("--output");

    cmd.assert().try_success().map(|assert| {
        let out = String::from_utf8_lossy(&assert.get_output().stdout);

        let mut settings = Settings::new();
        settings.add_script_timestamp_filter();

        settings.bind(|| {
            assert_snapshot!(out);
        });
    })?;

    temp_dir.close()?;

    Ok(())
}
