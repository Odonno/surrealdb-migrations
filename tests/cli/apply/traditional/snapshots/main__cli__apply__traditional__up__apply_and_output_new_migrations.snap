---
source: tests/cli/apply/traditional/up.rs
expression: out
snapshot_kind: text
---
-- Apply migration for AddAdminUser --


CREATE permission:create_post SET name = 'create_post';
CREATE permission:create_comment SET name = 'create_comment';

CREATE user:admin SET username = 'admin', email = 'admin@admin.fr', password = 'admin';
CREATE script_migration SET script_name = '[timestamp]_AddAdminUser';
-- Apply migration for AddPost --


CREATE post SET title = 'Hello world!', content = 'This is my first post!', author = user:admin;
CREATE script_migration SET script_name = '[timestamp]_AddPost';
-- Apply migration for CommentPost --


LET $post = (SELECT * FROM post LIMIT 1);

LET $comment = (RELATE user:admin->comment->$post SET content = "This is a comment on a post.");

RELATE user:admin->comment->$comment SET content = "This is a comment on a comment.";
CREATE script_migration SET script_name = '[timestamp]_CommentPost';
